### ISSUE TITLE
Unfulfillable nullable contraints are set for complex types with TPH entities

### ISSUE THREAD

**Issue by Wasserwecken** [dotnet/efcore#33004]
Usually, if a field has the `Required` attribute it is mapped as `nullable: false`.
But if this field is declared in a TPH entity, that condition may not be possible.

Unfortunately, required-flaged fields within complex types are set as NOT NULL anyway,
which prevents persisting entities that do not implement those fields.

![image](https://github.com/dotnet/efcore/assets/15609168/64f08b5c-6d8a-4b4d-a3e8-399c607446a6)

## Core for reproduction
Migrations are required (`Add-Migration Init -c CustomContext5`)

```C#
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Diagnostics;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace ConsoleApp5
{
    internal static class Program5
    {
        public static void Test(string[] args)
        {
            var supplier = new Supplier() { Name = "Supplier", Extra = new() { Field = "Field" } };
            var customer = new Customer() { Name = "Customer" };

            using (var ctx = new CustomContext5())
            {
                ctx.Database.Migrate();
                ctx.Add(supplier);
                ctx.SaveChanges();

                ctx.Add(customer);
                ctx.SaveChanges();
            }

            Console.WriteLine("Done!");
            Console.ReadLine();
        }

        public class CustomContext5 : DbContext
        {
            public DbSet<Contact> Contacts { get; set; }
            public DbSet<Supplier> Suppliers { get; set; }
            public DbSet<Customer> Customers { get; set; }

            protected override void OnConfiguring(DbContextOptionsBuilder builder)
            {
                builder.LogTo(Console.WriteLine, new[] { RelationalEventId.CommandExecuting }).EnableSensitiveDataLogging();
                builder.UseSqlServer(@"Server=(localdb)\mssqllocaldb;Database=TestDb;Trusted_Connection=True");
            }

            protected override void OnModelCreating(ModelBuilder builder)
            {
                builder.Entity<Contact>().HasDiscriminator(e => e.Discriminator);
            }
        }

        public abstract class Contact
        {
            public int Id { get; set; }
            public string? Discriminator { get; set; }
            public string? Name { get; set; }
        }

        public class Supplier : Contact
        {
            [Required]
            public string? AnotherExtra { get; set; }
            public required Extra Extra { get; set; }
        }

        public class Customer : Contact
        {
        }

        [ComplexType]
        public class Extra
        {
            [Required]
            public string? Field { get; set; }
        }
    }
}
```

### Include provider and version information
EF Core version: 8.0.1
Database provider: Microsoft.EntityFrameworkCore.SqlServer

**Comment by Wasserwecken** [comment]
__Minor note__: This also affects atomic field types such as `int` and `bool`.

**Comment by ajcvickers** [comment]
Note for team: still repros on latest daily. Patch candidate.

**Comment by maumar** [comment]
problem is in RelationalPropertyExtensions.IsColumnNullable - there we try to compute nullability of a column that the given property is mapped to. For inheritance scenario we look at declaring type and check if it's part of TPH. However for property of a complex type, declaring type is that complex type so we don't do the proper check. We should find the parent non-complex type instead

**Comment by roji** [comment]
Should we consider patching this (bug in new feature, extremely low risk)?

**Comment by maumar** [comment]
reopening for patch

**Comment by maumar** [comment]
merged to release/8.0, closing
